Feat: Add log-likelihood method for LGBN

yashnator opened the pull request:
Add method `log_likelihood(self, data)` to the LinearGaussianBayesianNetwork class

Example:

```python
from pgmpy.models import LinearGaussianBayesianNetwork
from pgmpy.factors.continuous import LinearGaussianCPD
import numpy as np
import pandas as pd

model = LinearGaussianBayesianNetwork([('x1', 'x2'), ('x2', 'x3')])
cpd1 = LinearGaussianCPD('x1', [1], 4)
cpd2 = LinearGaussianCPD('x2', [-5, 0.5], 4, ['x1'])
cpd3 = LinearGaussianCPD('x3', [4, -1], 3, ['x2'])
model.add_cpds(cpd1, cpd2, cpd3)
    
df = pd.DataFrame(np.random.normal(0, 1, (100, 3)), columns=['x1', 'x2', 'x3'])
ll = model.log_likelihood(df)

print("Log-likelihood: ", ll)
```

### Your checklist for this pull request
Please review the [guidelines for contributing](CONTRIBUTING.md) to this repository.

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!
- [x] Make sure you are making a pull request against the **dev branch** (left side). Also you should start *your branch* off *our dev*.
- [x] Check the commit's or even all commits' message styles matches our requested structure.

### Issue number(s) that this pull request fixes
- Fixes #2236

### List of changes to the codebase in this pull request
- Adds `log_likelihood(self, data)` method to `LinearGaussianBayesianNetwork` class

yashnator commented:
I was wondering if it is a good idea to benchmark the implementation with scipy's `multivariate_normal.logpdf` by converting to a joint Gaussian distribution.

Example: 

```python
import numpy as np
import pandas as pd
from pgmpy.models import LinearGaussianBayesianNetwork
from pgmpy.factors.continuous import LinearGaussianCPD
from scipy.stats import multivariate_normal

model = LinearGaussianBayesianNetwork([('x1', 'x2'), ('x2', 'x3')])
cpd1 = LinearGaussianCPD('x1', [1], 4)
cpd2 = LinearGaussianCPD('x2', [-5, 0.5], 4, ['x1'])
cpd3 = LinearGaussianCPD('x3', [4, -1], 3, ['x2'])
model.add_cpds(cpd1, cpd2, cpd3)

df = pd.DataFrame(np.random.normal(0, 1, (100, 3)), columns=['x1', 'x2', 'x3'])

mean, cov = model.to_joint_gaussian()

ll_joint = np.sum(multivariate_normal.logpdf(df.values, mean=mean, cov=cov))
ll_impl = model.log_likelihood(df)

print(f"LL from impl: {ll_impl}")
print(f"LL from scipy: {ll_joint}")
print(f"Difference: {abs(ll_impl - ll_joint)}")
print(f"Percentage error: {abs((ll_impl - ll_joint) / ll_joint) * 100}")
```

The error is within 1.5% after several runs. 

Should I make tests this way?

ankurankan commented:
@yashnator Thanks for the PR. My idea was to simply compute the joint distribution and get the loglikelihood from the multivariate Gaussian (what you have done in the comment above). Any reason to do it the way you have implemented?

yashnator commented:
I was trying to write code similar to the method from `bn_score`. I agree that using scipy is a better idea. Let me add some tests and rebase

yashnator committed:
Feat: Add log-likelihood method for LGBN

- Add method `log_likelihood(self, data)` to
the LinearGaussianBayesianNetwork class

Resolves: #2236

ankurankan commented:
@yashnator Let me know when this is ready for a review.

yashnator commented:
@ankurankan I have made the changes. Let me know if there are more changes

ankurankan commented on pgmpy/models/LinearGaussianBayesianNetwork.py:
```diff
@@ -263,6 +264,57 @@ def to_joint_gaussian(self) -> Tuple[np.ndarray, np.ndarray]:
+        Computes the log-likelihood of the given dataset under the current
+        Linear Gaussian Bayesian Network model.
+
+        Each variable is assumed to follow a Gaussian distribution conditioned
```
This needs to be changed?

ankurankan commented on pgmpy/models/LinearGaussianBayesianNetwork.py:
```diff
@@ -263,6 +264,57 @@ def to_joint_gaussian(self) -> Tuple[np.ndarray, np.ndarray]:
         # Round because numerical errors can lead to non-symmetric cov matrix.
         return mean.round(decimals=8), implied_cov.round(decimals=8)
 
+    def log_likelihood(self, data: Union[pd.DataFrame, np.ndarray]) -> float:
```
We can't accept `np.ndarray` here. Right? How will we know which columns corresponds to which variable.

yashnator replied:
Similar to the `score` method, I was thinking we could do `np.ndarray` as long as its columns are topologically sorted. However, I've removed `np.ndarray` as suggested

ankurankan commented on pgmpy/models/LinearGaussianBayesianNetwork.py:
```diff
@@ -263,6 +264,57 @@ def to_joint_gaussian(self) -> Tuple[np.ndarray, np.ndarray]:
+        data : pandas.DataFrame or array-like of shape (n_samples, n_features)
+            The dataset containing observations for all variables in the model.
+            If a DataFrame is provided, its columns must match the model's variable names.
+            If an array-like is provided, the variables must be ordered according to the model's topological order.
```
I think its simpler to just keep it to pandas data frame and not support numpy array.

ankurankan commented on pgmpy/models/LinearGaussianBayesianNetwork.py:
```diff
@@ -263,6 +264,57 @@ def to_joint_gaussian(self) -> Tuple[np.ndarray, np.ndarray]:
+        if isinstance(data, pd.DataFrame):
+            ordering = list(nx.topological_sort(self))
+            missing = set(ordering) - set(data.columns)
+            if missing:
```
We should try to keep all the input checks at the top of the method and not with the core logic.

ankurankan commented on pgmpy/tests/test_models/test_LinearGaussianBayesianNetwork.py:
```diff
@@ -393,6 +393,55 @@ def test_get_random(self):
+        try:
+            ll_good = self.model.log_likelihood(good_fit)
+            ll_bad = self.model.log_likelihood(bad_fit)
+        except Exception as e:
```
Why is this exception needed?

yashnator replied:
Used it for testing purposes. Removed.

ankurankan commented on pgmpy/tests/test_models/test_LinearGaussianBayesianNetwork.py:
```diff
@@ -393,6 +393,55 @@ def test_get_random(self):
+            "Expected higher log-likelihood for well-fitting data compared to random data.",
+        )
+
+        # Test log-likelihood with ndarray
```
I would suggest removing support for numpy array.

ankurankan reviewed (approved):
Thanks for the PR @yashnator .Looks good to me.
